include/deal.II/base/symmetric_tensor.h

```diff
@@ -2753,6 +2781,99 @@ invert (const SymmetricTensor<4,3,double> &t);
 
 
 /**
+ * Full contraction of three tensors: Return a scalar number that is the
+ * result of a full contraction of a symmetric tensor @p left of rank @p rank_1,
+ *  a tensor @p middle of rank $(\text{rank}_1+\text{rank}_2)$ and a tensor @p
+ * right of rank @p rank_2:
+ * @f[
+ *   \sum_{i_1,\ldots,i_{r1},j_1,\ldots,j_{r2}}
+ *   \text{left}_{i_1,\ldots,i_{r1}}
+ *   \text{middle}_{i_1,\ldots,i_{r1},j_1,\ldots,j_{r2}}
+ *   \text{right}_{j_1,\ldots,j_{r2}}
+ * @f]
+ *
+ * @relates Tensor
+ * @relates SymmetricTensor
+ * @author Jean-Paul Pelteret, Matthias Maier, 2017
+ */
+template <int rank_1, int rank_2, int dim,
+          typename T1, typename T2, typename T3>
+typename ProductType<T1, typename ProductType<T2, T3>::type>::type
+contract3 (const SymmetricTensor<rank_1, dim, T1> &left,
+           const Tensor<rank_1 + rank_2, dim, T2> &middle,
+           const Tensor<rank_2, dim, T3>          &right)
```

Comment:
I had thought about that, but discounted it to avoid ambiguity as to what constitutes a `Tensor` or `SymmetricTensor`. However, I'd be happy to make the change. I'll also make it clear in the documentation (and with a `@relates SymmetricTensor` tag) that this function, although defined in `tensor.h` is valid for both tensor types.